MVP

cofinley opened the pull request:
Includes abilities to alter config, setup presets to create notes from, word select, and more!

I believe it's at minimum viable product status based off what I would use it for and judging from basic needs. I realize there are some things we discussed which are not here yet (e.g. bulk create, non-exact search), but those can be added in the future and are not what I consider MVP at this time.

That being said, this isn't constituting a release, but at least a major step towards the finish line. The PR is mainly to hit the save button so it doesn't become an even bigger branch.

This is a big one, so I'd recommend testing before diving into the code.

### Video Preview

https://webmshare.com/play/qdve9

### Screenshots

Config, search settings:

![](https://i.imgur.com/c8edFxV.png)

Config, note creation presets:

![](https://i.imgur.com/VP0unP4.png)

Word Select:

![](https://i.imgur.com/OLfEleo.png)

Note creation based off preset in config:

![](https://i.imgur.com/oKUOMBB.png)

Note after creation, regular editor used:

![](https://i.imgur.com/Vj6DN7V.png)

cofinley committed:
WIP on word select window

Add flow layout and test case with documented data structure

cofinley committed:
Add UI for note creation window

- Two-pane window, left for word select, right for note editing/creation.
- Search for notes instead of cards, less redundancy.

cofinley committed:
Use stacked widget to browse current notes for performance

- Loading editors for all notes at one time was much too slow
    - Collapse and lazy-load with stacked widget
- Lazy-load one anki editor instance at a time for notes in QListWidget
- Reset note list on new word select
- Reset hooks on init

cofinley committed:
Reduce events, display current word

Hide note list if new word or if window just launched. Blank widgets
don't look great.

cofinley committed:
Clean up events, find context sentences, fix word select

Word select was listing the words out of a dictionary and therefore
was a set, not the original text as it was entered.

cofinley committed:
Start addon config in preparation for note creation presets

Create note field chooser like deckchooser for reusability
Allow note field tree to be single select (in prep. for single field
selection for note creation presets).

cofinley committed:
Flesh out config for default decks and notes/fields

Connect initial addon form to these config options successfully

cofinley committed:
WIP commit for config and note creation presets

cofinley committed:
Successfully save, load, rename, add, and remove note creation presets

Change preset data structure to indexed uuids for easier access and
manipulation.

cofinley committed:
Link note creation presets to the note creation pane

Next step is to use the preset to populate the add note form

cofinley committed:
Load selected word into add card dialog successfully

Now need to make the add cards dialog appear as a nested widget in the
note creation pane.

cofinley committed:
Convert anki's addcards dialog to widget, attempt mem. mgmt.

cofinley committed:
Allow sentences in note presets, fix sentence regex

Remove punctuation (except for single quote) on word select.

cofinley committed:
Fix bugs with dangling hooks

Fix overrides for close events for widgets so that hooks can be removed
and can be removed at the right time. Was getting latent errors about
objects which shouldn't have still had active hooks. Turns out the close()
method was the incorrect method to override, use closeEvent(event).

cofinley committed:
Fix bug with layout widget removal, refresh word select on note add

When removing widgets from a layout, we were looping on an index while
also popping from the end of the list, so half-way through the loop,
we would get None behavior. Loop backwards and use widget.setParent(None)
over widget.deleteLater().

This was needed to be able to properly refresh the word select pane
on note add. Now users will see a live update of the text when they add
notes with words found, and they are able to click on the now-found words
to see the notes they just created and edit them.

See https://stackoverflow.com/a/25330164/3203662

cofinley committed:
Refactor the clearing layouts and stack widgets

Use widget.deleteLater() instead of setParent(None) after a test:
https://pastebin.com/HzUrbrYC

setParent can still re-show the widget. deleteLater will not.

cofinley committed:
Improve UX of note presets in config, allow cancel in config

Add parent note type as a chooser for preset, only show applicable fields
in the word and sentence destination field choosers.

Only save config if user clicks "Save All."

cofinley committed:
Reorganize widgets into directories for each 'view.'

cofinley committed:
Choose sensible defaults on note presets

Choose different fields for word and sentence destinations, otherwise
blank field values. Do this on new preset and also on note type change.

cofinley committed:
Remove single selection mode for note field tree, fix config save error

Save config preset on config "Save All," even if no settings were changed.

Update config.md to reflect new note preset data structure.

cofinley committed:
Allow user to define a preset from note creation pane if no presets exist

cofinley committed:
Add documentation, disable ignored words until done

ohare93 commented:
Daaamn this looks slick :heart_eyes: I will have a play around with this tomorrow. Excellent job! :+1:

ohare93 commented:
Had a quick review of the functionality, though I haven't looked at any code. Need to go over this some more, but here are my initial thoughts I wrote down.

Love the word highlighting and them being buttons, I can really see this opening the door for full bucket functionality like we discussed. Though duplicate words still come through to the next view, and both are clickable. I guess that is intentional, just to bring through the whole paragraph as it is?

Does the "Define a note preset..." have to disappear when you already have note types setup? Just having a "+" button there would be good. Better yet, since the preset is still editable why not have a "Add current setup as preset" sort-of button?

When creating a note, I don't understand what the view underneath the "Create" buttons is for. Also I get a crash when I try to click on one of them, after agreeing to lose my current input:

![image](https://user-images.githubusercontent.com/1471350/66425983-281f6300-ea11-11e9-9195-ac86f4ee15a4.png)

```
Error 
An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed. 
If the issue only occurs when add-ons are enabled, please use the Tools>Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem. 
When you've discovered the add-on that is causing the problem, please report the issue on the add-ons section of our support site. 
Debug info:
Anki 2.1.15 (dev) Python 3.6.8 Qt 5.9.5 PyQt 5.10.1
Platform: Linux
Flags: frz=False ao=True sv=1
Add-ons possibly involved: findmissingwords

Caught exception:
Traceback (most recent call last):
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/search_results/note_creation.py", line 134, in display_note_editor
    note = self.mw.col.getNote(note_id)
  File "/home/jmo/Development/anki/anki/collection.py", line 263, in getNote
    return anki.notes.Note(self, id=id)
  File "/home/jmo/Development/anki/anki/notes.py", line 15, in __init__
    self.load()
  File "/home/jmo/Development/anki/anki/notes.py", line 38, in load
    from notes where id = ?""", self.id)
TypeError: 'NoneType' object is not iterable

```

After this the add-on is unresponsive, I cannot open the config options for instance, and need to restart Anki.

Regardless, excellent work my man :+1: I can see having QT to make the UI is really helping you get stuff setup and experiment! Need to get on that myself :sweat_smile:

cofinley commented:
Thanks for testing it out!

> Does the "Define a note preset..." have to disappear when you already have note types setup?

Good idea

>  I don't understand what the view underneath the "Create" buttons is for

Yeah those should be labeled

> I get a crash when I try to click on one of them, after agreeing to lose my current input

Interesting, the list item is supposed to be removed and therefore unclickable if you hit cancel and agree to lose input. I'll test this.

I'll work on these changes soon. Thanks again.

cofinley committed:
Fix bug with cancel not cancelling

Don't do anything if user denies cancel prompt from add notes widget
Keep button to create note presets always visible

cofinley commented:
Fixed the aforementioned issues (hopefully without introducing more bugs). Caught a bug where clicking "no" on cancel was just like clicking "yes."

ohare93 commented:
Still getting a crash, when the `self.note_ids list` is empty in `note_creation.py`

Here's my setup to get it: I simply put in those words, click a preset, then click the only "Notes Created" option in the list, it warns I will lose my current input and then I get
![image](https://user-images.githubusercontent.com/1471350/67042096-6ca0b200-f127-11e9-8123-74b4b5bc87ca.png)

I simply put in those words, click a preset, then click the only "Notes Created" option in the list, it warns I will lose my current input and then I get:

> Error 
> An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed. 
> If the issue only occurs when add-ons are enabled, please use the Tools>Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem. 
> When you've discovered the add-on that is causing the problem, please report the issue on the add-ons section of our support site. 
> Debug info:
> Anki 2.1.16 (dev) Python 3.6.8 Qt 5.9.5 PyQt 5.10.1
> Platform: Linux
> Flags: frz=False ao=True sv=1
> Add-ons possibly involved: findmissingwords
> 
> Caught exception:
> Traceback (most recent call last):
>   File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/search_results/note_creation.py", line 138, in display_note_editor
>     note_id = self.note_ids[index]
> IndexError: list index out of range
>

ohare93 commented:
> Ignored words (coming soon)

lol :100: :ok_hand:

ohare93 commented:
Apart from that bug, looks good, man! :ok_hand: Very nice work. Even with that bug, I am happy to pull this into main :grin:

jeprecated commented on src/find_missing_words/gui/config/__init__.py:
```diff
@@ -0,0 +1,37 @@
+from .note_creation_tab import NoteCreationTab
+
+
+class ConfigDialog(QDialog):
```
I thought there was some Python convention to do minimal work like loading 1 thing and kicking that off? :thinking: Have I merely imagined this? :sweat_smile:

cofinley replied:
Not sure I see what you mean? Lol 
Are you talking about this class?

jeprecated reviewed (approved):
God damn taking out the render methods _really_ clears up these add-ons! :open_mouth: it's so much nicer :grin: 

Looking good, man :+1:

cofinley commented:
I’ll take a look at that index out of range issue. Thanks!

cofinley committed:
Prevent redisplay error

Don't redisplay the same editor if clicking on the current selected note
list row. The action is redundant and results in null behavior.

cofinley committed:
Don't redisplay editor, but only when there is an editor shown

Otherwise, you can't click on a list item to display the editor.

Also, clear note editors on display_note_editor only when there is a
self.editor. The previous logic only did this when there wasn't an editor.

ohare93 commented:
Found another bug, where it crashes after clicking the "+" button.

Reproduction steps:
 1) Click on a word
 1) Click on a preset. You now have a populated word.
 1) Click on a word again (the same word, or a different one)
 1) The "+" button is now broken, no matter how you click around the window.

```
Error 
An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed. 
If the issue only occurs when add-ons are enabled, please use the Tools>Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem. 
When you've discovered the add-on that is causing the problem, please report the issue on the add-ons section of our support site. 
Debug info:
Anki 2.1.16 (dev) Python 3.6.8 Qt 5.9.5 PyQt 5.10.1
Platform: Linux
Flags: frz=False ao=True sv=1
Add-ons possibly involved: findmissingwords

Caught exception:
Traceback (most recent call last):
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/search_results/note_creation.py", line 194, in display_note_creation_config
    self.mw.find_missing_words_config = config_dialog = config.ConfigDialog(mw)
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/config/__init__.py", line 23, in __init__
    self.render_tabs()
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/config/__init__.py", line 27, in render_tabs
    self.note_creation_tab = NoteCreationTab(self.config)
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/config/note_creation_tab.py", line 29, in __init__
    self.render_note_creation_settings()
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/config/note_creation_tab.py", line 32, in render_note_creation_settings
    self.populate_presets()
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/config/note_creation_tab.py", line 45, in populate_presets
    note_preset_widget = note_preset.NotePreset(self.config, preset, self.update_preset)
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/config/note_preset.py", line 30, in __init__
    self.render_note_type_chooser()
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/config/note_preset.py", line 39, in render_note_type_chooser
    callback=self.update_note_type)
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/utils/list_chooser.py", line 18, in __init__
    self.update_button()
  File "/home/jmo/.local/share/Anki2/addons21/findmissingwords/gui/utils/list_chooser.py", line 50, in update_button
    self.mw.reset()
  File "/home/jmo/Development/anki/aqt/main.py", line 533, in reset
    runHook("reset")
  File "/home/jmo/Development/anki/anki/hooks.py", line 29, in runHook
    func(*args)
  File "/home/jmo/Development/anki/aqt/modelchooser.py", line 47, in onReset
    self.updateModels()
  File "/home/jmo/Development/anki/aqt/modelchooser.py", line 82, in updateModels
    self.models.setText(self.deck.models.current()['name'])
RuntimeError: wrapped C/C++ object of type QPushButton has been deleted

```

I assume the error is due to a lack of cleaning up the previous unfinished word when you click on a new word :sweat_smile:

cofinley committed:
Fix memory leak with AddNoteWidget

Display search filter on word select, otherwise users might be confused
why something is still green when they just added a note, but for a
different deck/model/field combo.

cofinley commented:
Fixed the above issue as well as another memory management bug. Going to merge.
